src/arithmetic/algebraic_expression.h

```diff
@@ -42,53 +46,30 @@ struct AlgebraicExpressionNode {
     AlgebraicExpressionNodeType type;
 };
 
-AlgebraicExpressionNode *AlgebraicExpressionNode_NewOperationNode(AL_EXP_OP op);
-AlgebraicExpressionNode *AlgebraicExpressionNode_NewOperandNode(GrB_Matrix operand);
+typedef struct {
+  AlgebraicExpressionNode *exp_root;
+  AlgebraicExpressionNode *exp_leaf;
+  Node *src;
+  Node *dest;
+  Edge *edge;
+} AE_Unit;
+
+AlgebraicExpressionNode* AlgebraicExpression_InvertTree(AlgebraicExpressionNode *root);
+AlgebraicExpressionNode* AlgebraicExpression_Append(AlgebraicExpressionNode *root, AlgebraicExpressionNode *child);
+int AlgebraicExpression_OperandCount(AlgebraicExpressionNode *root);
+AlgebraicExpressionNode* AlgebraicExpression_PopFirst(AlgebraicExpressionNode **root);
+AlgebraicExpressionNode* AlgebraicExpression_PopLast(AlgebraicExpressionNode **root);
+AE_Unit*** AlgebraicExpression_BuildExps(const AST *ast, const QueryGraph *q, Node **starting_points, int component_count);
+
+AlgebraicExpressionNode* AlgebraicExpression_NewOperationNode(AL_EXP_OP op);
+AlgebraicExpressionNode* AlgebraicExpression_NewOperand(GrB_Matrix mat);
+AlgebraicExpressionNode* AlgebraicExpression_NewNodeOperand(Node *operand);
+AlgebraicExpressionNode* AlgebraicExpression_NewEdgeOperand(Edge *operand);
 void AlgebraicExpressionNode_AppendLeftChild(AlgebraicExpressionNode *root, AlgebraicExpressionNode *child);
 void AlgebraicExpressionNode_AppendRightChild(AlgebraicExpressionNode *root, AlgebraicExpressionNode *child);
 void AlgebraicExpression_SumOfMul(AlgebraicExpressionNode **root);
+void AlgebraicExpression_EvalWithFilter(AlgebraicExpressionNode *exp, GrB_Matrix filter, GrB_Matrix res);
```

Comment:
When a filter vector/matrix is to be applied, simply append that filter with a multiplication operation to the tree.
Consider the simplest tree, just a single operand, when a filter matrix is applied to it the transformed tree becomes:
(original operand)<-ROOT_OP(*)->(filter matrix)

Reply:
Right now, the actual behavior is just like you describe. Since the Eval gets called once for every input column, though, I'd ultimately like to introduce the filter matrix in a more efficient way than traversing the tree to append it, then traversing again to perform the series of operations. This approach is less sensible for trees than it was for our previous array-based AlgebraicExpression.

Reply:
I think we only need to modify the tree once,
and then updated the appended matrix/vector for every incoming set of source nodes.

Reply:
:+1: Good call, that makes sense!